iocontrol/speedgoat/controller.go

```diff
@@ -1,45 +1,198 @@
+	if err != nil {
+		return errors.Wrap(err, "dial speedgoat")
+	}
+
+	c.conn = conn
+	c.opened = true
+
+	go c.tickOutputs()
+	go c.tickInputs()
+
 	return nil
 }
 
-// SetDigital sets an output digital pin for a Speedgoat digital pin
-func (c *Controller) SetDigital(output *DigitalPin, b bool) error {
+// Close closes any resources related to the controller.
+func (c *Controller) Close() error {
+	c.l.Info("closing speedgoat controller")
+
+	c.opened = false
+
+	err := c.conn.Close()
+	if err != nil {
+		return errors.Wrap(err, "close speedgoat connection")
+	}
 	return nil
 }
 
-// ReadDigital returns the level of a Speedgoat digital pin
-func (c *Controller) ReadDigital(output *DigitalPin) (bool, error) {
-	return false, nil
+// SetDigital sets an output digital pin for a Speedgoat digital pin.
+func (c *Controller) SetDigital(output *DigitalPin, b bool) {
+	c.muDigital.Lock()
+	defer c.muDigital.Unlock()
+	c.digital[output.Index] = b
 }
 
-// WriteVoltage sets the voltage of a Speedgoat analog pin
-func (c *Controller) WriteVoltage(output *AnalogPin, voltage float64) error {
-	return nil
+// ReadDigital returns the level of a Speedgoat digital pin.
+func (c *Controller) ReadDigital(output *DigitalPin) bool {
+	c.muDigital.Lock()
+	defer c.muDigital.Unlock()
+	return c.digital[output.Index]
 }
 
-// ReadVoltage returns the voltage of a Speedgoat analog pin
-func (c *Controller) ReadVoltage(output *AnalogPin) (float64, error) {
-	return 0.00, nil
+// WriteVoltage sets the voltage of a Speedgoat analog pin.
+func (c *Controller) WriteVoltage(output *AnalogPin, voltage float64) {
+	c.muAnalog.Lock()
+	defer c.muDigital.Unlock()
+	c.analog[output.Index] = voltage
+}
+
+// ReadVoltage returns the voltage of a Speedgoat analog pin.
+func (c *Controller) ReadVoltage(output *AnalogPin) float64 {
+	c.muAnalog.Lock()
+	defer c.muDigital.Unlock()
+	return c.analog[output.Index]
 }
 
-// WriteCurrent sets the current of a Speedgoat analog pin
+// WriteCurrent sets the current of a Speedgoat analog pin (unimplemented for Speedgoat).
 func (c *Controller) WriteCurrent(output *AnalogPin, current float64) error {
 	return nil
 }
 
-// ReadCurrent returns the current of a Speedgoat analog pin
+// ReadCurrent returns the current of a Speedgoat analog pin (unimplemented for Speedgoat).
 func (c *Controller) ReadCurrent(output *AnalogPin) (float64, error) {
 	return 0.00, nil
 }
+
+// tickOutputs transmits the packed data for the digital and analog outputs to the Speedgoat at a set time interval.
+func (c *Controller) tickOutputs() {
+	ticker := time.NewTicker(_tickTime)
+	defer ticker.Stop()
+
+	for c.opened {
+		for range ticker.C {
+			_, err := c.conn.Write(c.packOutputs())
+			if err != nil {
+				c.l.Error("speedgoat controller", zap.Error(errors.Wrap(err, "connection write")))
+			}
+		}
+	}
+}
+
+// tickInputs reads the pin data from the connection and unpacks it into its respective pin arrays.
+func (c *Controller) tickInputs() {
```

Comment:
tickInputs and tickOutputs have same frequency. Can this lead to race conditions if we try to read and write same time?